tests/test_authority.py

```diff
@@ -104,32 +104,63 @@ def test_authority_with_path_should_be_used_as_is(self, oidc_discovery):
     "authorization_endpoint": "https://contoso.com/authorize",
     "token_endpoint": "https://contoso.com/token",
     })
-class TestOidcAuthority(unittest.TestCase):
+class OidcAuthorityTestCase(unittest.TestCase):
+    authority = "https://contoso.com/tenant"
+
+    def setUp(self):
+        # setUp() gives subclass a dynamic setup based on their authority
+        self.oidc_discovery_endpoint = (
+            # MSAL Python always does OIDC Discovery,
+            # not to be confused with Instance Discovery
+            # Here the test is to confirm the OIDC endpoint contains no "/v2.0"
+            self.authority + "/.well-known/openid-configuration")
+
     def test_authority_obj_should_do_oidc_discovery_and_skip_instance_discovery(
             self, oidc_discovery, instance_discovery):
         c = MinimalHttpClient()
-        a = Authority(None, c, oidc_authority_url="https://contoso.com/tenant")
+        a = Authority(None, c, oidc_authority_url=self.authority)
         instance_discovery.assert_not_called()
-        oidc_discovery.assert_called_once_with(
-            "https://contoso.com/tenant/.well-known/openid-configuration", c)
+        oidc_discovery.assert_called_once_with(self.oidc_discovery_endpoint, c)
         self.assertEqual(a.authorization_endpoint, 'https://contoso.com/authorize')
         self.assertEqual(a.token_endpoint, 'https://contoso.com/token')
 
     def test_application_obj_should_do_oidc_discovery_and_skip_instance_discovery(
             self, oidc_discovery, instance_discovery):
         app = msal.ClientApplication(
-            "id",
-            authority=None,
-            oidc_authority="https://contoso.com/tenant",
-            )
+            "id", authority=None, oidc_authority=self.authority)
         instance_discovery.assert_not_called()
         oidc_discovery.assert_called_once_with(
-            "https://contoso.com/tenant/.well-known/openid-configuration",
-            app.http_client)
+            self.oidc_discovery_endpoint, app.http_client)
         self.assertEqual(
             app.authority.authorization_endpoint, 'https://contoso.com/authorize')
         self.assertEqual(app.authority.token_endpoint, 'https://contoso.com/token')
 
+
+class DstsAuthorityTestCase(OidcAuthorityTestCase):
+    # Inherits OidcAuthority's test cases and run them with a dSTS authority
+    authority = (  # dSTS is single tenanted with a dummy tenant placeholder
+        'https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/tid')
```

Comment:
Can we also add a test case with common? I think with dsts there can be a tenant id or common used in the authority.

Reply:
> I'm not really sure how _common_ is "common" :) and user discovery in dsts. It may be worth asking Sai before.

You meant OIDC discovery? This is the [OIDC discovery for common](https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/common/.well-known/openid-configuration) and this is for [the non-common](https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/contoso/.well-known/openid-configuration). I'll get back to you when I hear back from Sai.

Reply:
Right, "common" is supported but doesn't have the same meaning and eSTS "common", since dSTS is single tenanted. So to me it doesn't make sense to inherit all of eSTS complex logic around multi-tenancy, such as replacing "common" with the id token's `tid` in the cache etc. 

A test is fine.

Reply:
> You meant OIDC discovery? This is the [OIDC discovery for common](https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/common/.well-known/openid-configuration) and this is for [the non-common](https://test-instance1-dsts.dsts.core.azure-test.net/dstsv2/contoso/.well-known/openid-configuration). I'll get back to you when I hear back from Sai.

Confirmed with Sai that the seemingly two sets of authorities, common and un-common, are just a side effect of reusing some eSTS code. Under the hood, that common endpoint will still convert the "common" into the tenant id and mint a token for that tenant id.

> So to me it doesn't make sense to inherit all of eSTS complex logic around multi-tenancy, such as replacing "common" with the id token's `tid` in the cache etc.

MSAL Python's [tenant agnostic approach](https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/341#issuecomment-1400936207) does not attempt converting "common" into tid.